test/pytest/test_bridge.py

```diff
@@ -419,3 +421,52 @@ async def test_freeze(self):
         await self.transport.assert_data(koelle, b'x3')
         await self.transport.assert_msg('', command='done', channel=koelle)
         await self.transport.assert_msg('', command='close', channel=koelle)
+
+    async def test_peer_routing_rule(self):
+        await self.start()
+
+        # inject a custom rule
+        args, env = cockpit.superuser.SUPERUSER_BRIDGES['pseudo']
+        pseudo_rule = cockpit.peer.PeerRoutingRule(self.bridge, args, env, lambda msg: msg.get('for-pseudo'))
+        self.bridge.routing_rules.insert(0, pseudo_rule)
+
+        # verify that a normal message gets routed normally
+        echo = await self.transport.check_open('echo')
+        assert isinstance(self.bridge.open_channels[echo], cockpit.channels.EchoChannel)
+
+        # verify that a for-pseudo message gets correctly routed.  split the
+        # open into two parts so we can observe the two states.
+        pseudo_echo = self.transport.send_open('echo', for_pseudo=True)
+        peer = self.bridge.open_channels[pseudo_echo]
+        assert isinstance(peer, cockpit.peer.Peer)
+        assert peer.name == args[0]
+
+        # We opened the channel, which should have started spawning the
+        # program, but we've yet to return to the mainloop, so there is no way
+        # that we should have been able to receive init yet.
+        # Make sure the peer is frozen and the open message is in the queue.
+        assert peer.queue is not None
+        method, (channel, command, message) = peer.queue[0]
+        assert method.__self__ == peer
+        assert method.__name__ == 'do_channel_control'
+        assert channel == pseudo_echo
+        assert command == 'open'
+        assert message == {'command': 'open', 'channel': pseudo_echo, 'payload': 'echo', 'for-pseudo': True}
+
+        # wait for the open reply, which should allow us to receive init and thaw the queue
+        await self.transport.assert_msg('', command='ready', channel=pseudo_echo)
+        assert peer.queue is None
+
+        # and of course, it should echo
+        self.transport.send_data(pseudo_echo, b'hihi\n')
+        await self.transport.assert_data(pseudo_echo, b'hihi\n')
```

Comment:
It's not a surprise that `pseudo_echo` replies with `hihi` -- it's a standard "echo" channel after all, i.e. at least that part of the test would succeed just equally well without the `for_pseudo`. Could this somehow test that "hihi" actually gets sent to and processed by the executed peer program? Or even better, could this change test/pytest/pseudo.py to do something different than just echoing, like returning the reverse of its input?